### DISCUSSION TITLE
Order of `BaseModel.model_fields`

### DISCUSSION THREAD

**Discussion by kasbah** [pydantic/pydantic#8289] [Question] [answered]
I am migrating a project from using dataclassess to v2.5 `BaseModel`. We have been using `dataclasses.fields` to iterate through fields in a pre-defined order.

So from:

```python
for field in dataclasses.fields(MyClass):
    name = field.name
    # we may break or continue here so the order matters
    ...
```
to:

```python
for name, field in MyClass.model_fields.items():
    # we may break or continue here so the order matters
    ...
```
 Is the ordering of `BaseModel.model_fields.items()` guaranteed? It not being an `OrderedDict` but a `dict` makes me think not but then again seems like lots of internal Pydantic things depend on the order of fields too, how do they do it?

**Comment by uriyyo** [comment, marked as answer]
Hi @kasbah,

You can checkout this article - https://medium.com/junior-dev/python-dictionaries-are-ordered-now-but-how-and-why-5d5a40ee327f

Basically, starting from python version 3.7 dictionary order is guaranteed to be insertion order.

**Reply by kasbah** [reply]
Oh, nice, thanks!

**Reply by loucadufault** [reply]
The natural follow up question: what is pydantic's insertion order? Is it the order that the properties are defined in the model class? or the order of the args/kwargs to the model constructor?

**Comment by g-rutter** [comment]
As the ordering of Python dict is not guaranteed in the spec but **incidentally** arises from the implementation, it would technically be better to have guaranteed way to iterate the fields in order.
